aips/aip-xx-enums.md

```diff
@@ -0,0 +1,253 @@
+A number of new instructions need to be introduced to deal with enums, notably a new form of borrow operation on fields of variants that aborts for unexpected variants. The later one is needed for allowing frontends to generate matching code via references to the data.
+
+## Impact
+
+This is a significant extension to the Move VM which will highly improve the fidelity of the code which can be written in Move. This first version contains the minimal necessary set of new features, more are needed in the future for improving performance.
+
+## Alternative solutions
+
+There are no viable alternatives known. The closest would be to represent a variant type as a struct with optional fields, but this is inefficient and would not give upgradability. 
+
+## Specification and Implementation Details
+
+### Representation of Enum Types
+
+To represent enum types (aka 'variant structs') the existing `StructDefinition` type is extended as follows:
+
+```rust
+pub struct StructDefinition {
+    ...
+    pub field_information: StructFieldInformation,
+}
+pub enum StructFieldInformation {
+    Native,
+    Declared(Vec<FieldDefinition>),
+    // New variant
+    DeclaredVariants(Vec<VariantDefinition>),
+}
+pub struct FieldDefinition {
+    pub name: IdentifierIndex,
+    pub signature: TypeSignature,
+}
+// New struct
+pub struct VariantDefinition {
+    pub name: IdentifierIndex,
+    pub fields: Vec<FieldDefinition>,
+}
+```
+
+The extension is neutral in serialization size.
+
+
+### Representation as VM 1.0 Value Types
+
+In the current Move VM value system, an enum is represented as a container of values (similar to structs and vectors). The first element in this container is a `u16` with the variant tag, the remaining values are interpreted depending on the variant.
+
+### Serialization
+
+The serialization format starts with the `u16` variant tag, followed by the values of the fields of that variant. During type-driven deserialization, first the variant tag is read, then the according fields with types derived from the chosen variant.
+
+### Upgrade Compatibility
+
+Variant structs can be upgraded by adding new variants. The new variants must be added, in sequence, after the existing ones which must be a prefix of the variants of the new variant struct.
+
+### Handles
+
+A number of new handle types are needed to access variant structs from instructions. The general principle of Move bytecode is that arguments to an instruction are represented as indices in tables of such handles.
+
+First a givem variant of a struct needs a denotation for operations like pack, unpack, and variant testing:
+
+```rust
+pub struct StructVariantHandle {
+    pub struct_index: StructDefinitionIndex,
+    pub variant: VariantIndex,
+}
+
+```
+
+Second, since field borrowing for variant structs is supported, a way to denote a field in a variant is required:
+
+```rust
+pub struct VariantFieldHandle {
+    pub owner: StructDefinitionIndex,
+    pub variants: Vec<VariantIndex>,
```

Comment:
VariantIndex is not defined here, would be useful to have.

does index in this vector mean anything, or is this just a "set" of viable variants? how does this handle if field is different for different variant? (i.e. if same name of the field has different positions in different variants)

if this is set would it be better to have it as HashSet or bit vector?

Reply:
Added definition of VariantIndex, and some more clarifying text